File: problems/树/105.从前序与中序遍历序列构造二叉树.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <functional>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
    unordered_map<int, int> idx;
    for (int i = 0; i < (int)inorder.size(); i++) idx[inorder[i]] = i;
    function<TreeNode*(int,int,int,int)> build = [&](int pl, int pr, int il, int ir) -> TreeNode* {
        if (pl > pr) return nullptr;
        TreeNode* root = new TreeNode(preorder[pl]);
        int mid = idx[preorder[pl]];
        root->left  = build(pl+1, pl+mid-il, il, mid-1);
        root->right = build(pl+mid-il+1, pr, mid+1, ir);
        return root;
    };
    return build(0, (int)preorder.size()-1, 0, (int)inorder.size()-1);
}
void inorderPrint(TreeNode* root) {
    if (!root) return;
    inorderPrint(root->left);
    cout << root->val << " ";
    inorderPrint(root->right);
}
int main() {
    vector<int> pre1 = {3, 9, 20, 15, 7};
    vector<int> in1  = {9, 3, 15, 20, 7};
    inorderPrint(buildTree(pre1, in1)); cout << endl;  // 9 3 15 20 7
    vector<int> pre2 = {1};
    vector<int> in2  = {1};
    inorderPrint(buildTree(pre2, in2)); cout << endl;  // 1
    vector<int> pre3, in3;
    cout << (buildTree(pre3, in3) == nullptr ? "null" : "?") << endl;  // null
    return 0;
}

File: problems/树/105.从前序与中序遍历序列构造二叉树.py
#!/usr/bin/env python3
"""
105. 从前序与中序遍历序列构造二叉树
"""
from typing import List, Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def build_tree(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    if not preorder: return None
    root = TreeNode(preorder[0])
    idx = inorder.index(preorder[0])
    root.left  = build_tree(preorder[1:idx+1], inorder[:idx])
    root.right = build_tree(preorder[idx+1:], inorder[idx+1:])
    return root

def inorder_list(root: Optional[TreeNode]) -> List[int]:
    if not root: return []
    return inorder_list(root.left) + [root.val] + inorder_list(root.right)

if __name__ == "__main__":
    preorder1 = [3, 9, 20, 15, 7]
    inorder1  = [9, 3, 15, 20, 7]
    root1 = build_tree(preorder1, inorder1)
    print(inorder_list(root1))  # [9, 3, 15, 20, 7]
    print(inorder_list(build_tree([1], [1])))  # [1]
    print(build_tree([], []))  # None
